File: HWW_polarization/JHUGen_PolarizationModel/GenMELA_WW.py

import ROOT
import uproot
import numpy as np
import pandas as pd
import mplhep as hep
import matplotlib.pyplot as plt
import subprocess

import argparse
import sys
import os
import time
import json
from pathlib import Path

ROOT.gROOT.SetBatch(True)
ROOT.TH1.SetDefaultSumw2(True)
ROOT.DisableImplicitMT()

def defaultParser():
    parser = argparse.ArgumentParser(add_help=False)

    def list_of_strings(arg):
        return arg.split(',')

    parser.add_argument(
        "-i",
        "--inDir",
        type=str,
        help="prefix for the job input",
        required=False,
        default="",
    )

    return parser

def createInputBranches(inFile):

    ####################
    #################### Code to test the gen. level polarized ME reweight with JHUGen / MELA
    ####################
    
    inTree = "Events"
    outFile = os.environ["TMPDIR"] + "/" + inFile.split("/")[-1]
    
    ROOT.gSystem.Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MELACalc/JHUGen-v757.3/JHUGenMELA/MELA/data/el9_amd64_gcc12/libmcfm_711.so","", ROOT.kTRUE)
    ROOT.gSystem.Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MELACalc/JHUGen-v757.3/JHUGenMELA/MELA/data/el9_amd64_gcc12/libJHUGenMELAMELA.so","", ROOT.kTRUE)
    ROOT.gSystem.Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/IvyFramework/IvyDataTools/lib/libIvyFrameworkIvyDataTools.so","", ROOT.kTRUE)
    ROOT.gSystem.Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/IvyFramework/IvyAutoMELA/lib/libIvyFrameworkIvyAutoMELA.so","", ROOT.kTRUE)
    ROOT.gSystem.Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MelaAnalytics/GenericMEComputer/lib/libMelaAnalyticsGenericMEComputer.so","", ROOT.kTRUE)
    ROOT.gSystem.Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MelaAnalytics/EventContainer/lib/libMelaAnalyticsEventContainer.so","", ROOT.kTRUE)
    ROOT.gSystem.Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MelaAnalytics/CandidateLOCaster/lib/libMelaAnalyticsCandidateLOCaster.so","", ROOT.kTRUE)
    
    # Compile
    
    #gSystem->Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MELACalc/JHUGen-v757.3/JHUGenMELA/MELA/data/el9_amd64_gcc12/libmcfm_711.so", "", true);
    #gSystem->Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MELACalc/JHUGen-v757.3/JHUGenMELA/MELA/data/el9_amd64_gcc12/libJHUGenMELAMELA.so", "", true);
    #gSystem->Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/IvyFramework/IvyDataTools/lib/libIvyFrameworkIvyDataTools.so", "", true);
    #gSystem->Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/IvyFramework/IvyAutoMELA/lib/libIvyFrameworkIvyAutoMELA.so", "", true);
    #gSystem->Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MelaAnalytics/GenericMEComputer/lib/libMelaAnalyticsGenericMEComputer.so", "", true);
    #gSystem->Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MelaAnalytics/EventContainer/lib/libMelaAnalyticsEventContainer.so", "", true);
    #gSystem->Load("/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MelaAnalytics/CandidateLOCaster/lib/libMelaAnalyticsCandidateLOCaster.so", "", true);
    
    # -----
    
    ROOT.gSystem.Load("/afs/cern.ch/work/s/sblancof/private/Run2Analysis/sendEOSJobs/JHUGen_PolarizationModel/GenMELA_WW_cc.so","", ROOT.kTRUE)
    #ROOT.gSystem.Load("/afs/cern.ch/work/s/sblancof/private/Run2Analysis/sendEOSJobs/JHUGen_PolarizationModel/GenMELA_WW_2016_cc.so","", ROOT.kTRUE)
    
    ROOT.gInterpreter.Declare("GEN_POLARIZATION b(1);")
    #ROOT.gInterpreter.Declare("GEN_POLARIZATION_2016 b(1);")
    
    df = ROOT.RDataFrame(inTree, inFile)
    df = df.Define(
        "MELAOutput",
        "b(nLHEPart,LHEPart_pt,LHEPart_eta,LHEPart_phi,LHEPart_mass,LHEPart_incomingpz,LHEPart_pdgId,LHEPart_status,LHEPart_spin,GenPart_genPartIdxMother,GenPart_pdgId,GenPart_status,GenPart_pt,GenPart_eta,GenPart_phi,GenPart_mass,Generator_x1,Generator_x2,Generator_id1,Generator_id2)"
    )
    
    df = df.Define("LHEDaughterId", "MELAOutput[0]")
    df = df.Define("LHEDaughterPt", "MELAOutput[1]")
    df = df.Define("LHEDaughterEta", "MELAOutput[2]")
    df = df.Define("LHEDaughterPhi", "MELAOutput[3]")
    df = df.Define("LHEDaughterMass", "MELAOutput[4]")
    df = df.Define("LHEAssociatedParticlePt", "MELAOutput[5]")
    df = df.Define("LHEAssociatedParticleEta", "MELAOutput[6]")
    df = df.Define("LHEAssociatedParticlePhi", "MELAOutput[7]")
    df = df.Define("LHEAssociatedParticleMass", "MELAOutput[8]")
    df = df.Define("LHEAssociatedParticleId", "MELAOutput[9]")
    df = df.Define("LHEMotherId", "MELAOutput[10]")
    df = df.Define("LHEMotherPz", "MELAOutput[11]")
    df = df.Define("LHEMotherE", "MELAOutput[12]")
    
    df = df.Define("otherVars", "MELAOutput[13]")
    
    df = df.Define("qH", "otherVars[0]")
    df = df.Define("mV1", "otherVars[1]")
    df = df.Define("mV2", "otherVars[2]") 
    df = df.Define("costheta1", "otherVars[3]")
    df = df.Define("costheta2", "otherVars[4]")
    df = df.Define("Phi", "otherVars[5]")
    df = df.Define("costhetastar", "otherVars[6]")
    df = df.Define("Phi1", "otherVars[7]")
    df = df.Define("genMass", "otherVars[8]")
    
    columnsToStore = [
        "LHEDaughterId",
        "LHEDaughterPt",
        "LHEDaughterEta",
        "LHEDaughterPhi",
        "LHEDaughterMass",
        "LHEAssociatedParticlePt",
        "LHEAssociatedParticleEta",
        "LHEAssociatedParticlePhi",
        "LHEAssociatedParticleMass",
        "LHEAssociatedParticleId",
        "LHEMotherId",
        "LHEMotherPz",
        "LHEMotherE",
        "qH",
        "mV1",
        "mV2",
        "costheta1",
        "costheta2",
        "Phi",
        "costhetastar",
        "Phi1",
        "genMass",
        #"mll",
        #"dphill",
        #"XSWeight"
    ]
    
    print("Making a snapshot")

    opts = ROOT.RDF.RSnapshotOptions()
    #opts.fLazy = True
    opts.fMode = "UPDATE"
    opts.fOverwriteIfExists = True
    opts.fCompressionAlgorithm = ROOT.ROOT.kLZMA
    opts.fCompressionLevel = 9
    
    df.Snapshot("Events", outFile, columnsToStore, opts) # RECREATE
    print("DONE.")


def main():
    parser = defaultParser()
    args = parser.parse_args()

    inFile = args.inDir

    print("Run MELA variable builder -------")

    createInputBranches(inFile)


if __name__ == '__main__':
    main()
    print("DONE!")
    
"""
df = df.Define("weight", "XSWeight * 137.54")
df = df.Define("polWeight", "MELAOutput[0] * weight")
df = df.Define("qH", "MELAOutput[1]")
df = df.Define("mV1", "MELAOutput[2]")
df = df.Define("mV2", "MELAOutput[3]")
df = df.Define("costheta1", "MELAOutput[4]")
df = df.Define("costheta2", "MELAOutput[5]")
df = df.Define("Phi", "MELAOutput[6]")
df = df.Define("costhetastar", "MELAOutput[7]")
df = df.Define("Phi1", "MELAOutput[8]")
df = df.Define("genMass", "MELAOutput[9]")

h_mV1 = df.Histo1D(("hist_mV1", "hist_mV1", 64, 70., 130.), "mV1", "weight")
h_mV1_pol = df.Histo1D(("hist_mV1_pol", "hist_mV1_pol", 64, 70., 130.), "mV1", "polWeight")

h_mV2 = df.Histo1D(("hist_mV2", "hist_mV2", 64, 0., 70.), "mV2", "weight")
h_mV2_pol = df.Histo1D(("hist_mV2_pol", "hist_mV2_pol", 64, 0., 70.), "mV2", "polWeight")

h_genMass = df.Histo1D(("hist_genMass", "hist_genMass", 64, 90., 200.), "genMass", "weight")
h_genMass_pol = df.Histo1D(("hist_genMass_pol", "hist_genMass_pol", 64, 90., 200.), "genMass", "polWeight")

h_costheta1 = df.Histo1D(("hist_costheta1", "hist_costheta1", 64, -1., 1.), "costheta1", "weight")
h_costheta1_pol = df.Histo1D(("hist_costheta1_pol", "hist_costheta1_pol", 64, -1., 1.), "costheta1", "polWeight")

h_costheta2 = df.Histo1D(("hist_costheta2", "hist_costheta2", 64, -1., 1.), "costheta2", "weight")
h_costheta2_pol = df.Histo1D(("hist_costheta2_pol", "hist_costheta2_pol", 64, -1., 1.), "costheta2", "polWeight")

h_costhetastar = df.Histo1D(("hist_costhetastar", "hist_costhetastar", 64, -1., 1.), "costhetastar", "weight")
h_costhetastar_pol = df.Histo1D(("hist_costhetastar_pol", "hist_costhetastar_pol", 64, -1., 1.), "costhetastar", "polWeight")

h_Phi = df.Histo1D(("hist_Phi", "hist_Phi", 64, -3.14, 3.14), "Phi", "weight")
h_Phi_pol = df.Histo1D(("hist_Phi_pol", "hist_Phi_pol", 64, -3.14, 3.14), "Phi", "polWeight")

tfile = ROOT.TFile.Open("JHUGen_reweight_histograms.root", "RECREATE")
tfile.cd()
h_mV1.Clone().Write()
h_mV1_pol.Clone().Write()
h_mV2.Clone().Write()
h_mV2_pol.Clone().Write()
h_genMass.Clone().Write()
h_genMass_pol.Clone().Write()
h_costheta1.Clone().Write()
h_costheta1_pol.Clone().Write()
h_costheta2.Clone().Write()
h_costheta2_pol.Clone().Write()
h_costhetastar.Clone().Write()
h_costhetastar_pol.Clone().Write()
h_Phi.Clone().Write()
h_Phi_pol.Clone().Write()
tfile.Close()
"""

File: HWW_polarization/JHUGen_PolarizationModel/GenMELA_WW.cc
#include "TLorentzVector.h"

#include <iostream>
#include "ROOT/RVec.hxx"

#include "/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MelaAnalytics/EventContainer/interface/MELAEvent.h"
#include "/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/IvyFramework/IvyAutoMELA/interface/IvyMELAHelpers.h"
#include "/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/JHUGenMELA/MELA/interface/PDGHelpers.h"
#include "/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/JHUGenMELA/MELA/interface/TUtil.hh"
#include "/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MelaAnalytics/EventContainer/interface/HiggsComparators.h"
#include "/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MelaAnalytics/EventContainer/interface/TopComparators.h"
#include "/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/MelaAnalytics/CandidateLOCaster/interface/MELACandidateRecaster.h"
#include "/eos/user/s/sblancof/Run2Analysis/mkShapesRDF/JHUGenMELA/MELA/interface/Mela.h"

#include <vector>
#include <string>
#include <cstdio>
#include <cmath>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

#include "Mela.h"
#include "ZZMatrixElement.h"
#include "VectorPdfFactory.h"
#include "TensorPdfFactory.h"
#include "RooqqZZ_JHU_ZgammaZZ_fast.h"
#include "RooqqZZ_JHU.h"
#include "SuperMELA.h"
#include "TUtilHelpers.hh"
#include "MELAStreamHelpers.hh"

#include "RooMsgService.h"
#include "TFile.h"
#include "TH1F.h"
#include "TH2F.h"
#include "TH3F.h"
#include "TGraph.h"
#include "TSpline.h"
#include "TString.h"

using namespace RooFit;
using MELAStreamHelpers::MELAout;
using MELAStreamHelpers::MELAerr;

using namespace ROOT;
using namespace ROOT::VecOps;
using namespace std;
using LorentzVectorM = ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiM4D<float>>;


class GEN_POLARIZATION{
public:
  
  int nLHEPart;
  RVecF LHEPart_pt;
  RVecF LHEPart_eta;
  RVecF LHEPart_phi;
  RVecF LHEPart_mass;
  RVecI LHEPart_incomingpz;
  RVecI LHEPart_pdgId;
  RVecI LHEPart_status;
  RVecI LHEPart_spin;

  RVecI GenPart_genPartIdxMother;
  RVecI GenPart_pdgId;
  RVecI GenPart_status;
  RVecF GenPart_pt;
  RVecF GenPart_eta;
  RVecF GenPart_phi;
  RVecF GenPart_mass;

  float Generator_x1;
  float Generator_x2;
  int Generator_id1;
  int Generator_id2;

  bool recast();
  
  MELAEvent::CandidateVVMode VVMode;
  bool isGen;
  int VVDecayMode;

  double LHCsqrts;
  double mh;
  TVar::VerbosityLevel myVerbosity_;
  Mela* mela;

  float M_Z = 91.1876;
  float M_W = 80.399;
  float M_H = 125.0;
  float Gf = 1.16639e-5;
  float vev = 1.0 / sqrt(Gf*sqrt(2.0));
  float gwsq = 4.0 * M_W*M_W / (vev*vev);  
  
  float a1 = M_Z*M_Z/(vev*vev);
  float a2;
  float fL; 
  
  GEN_POLARIZATION(float f_L){

    VVMode = MELAEvent::getCandidateVVModeFromString("WW");
    isGen = true;
    VVDecayMode = 0; // For both ggH and VBF

    LHCsqrts = 13.6;
    mh = 125.0;
    myVerbosity_ = TVar::SILENT; //TVar::DEBUG; TVar::SILENT

    mela = new Mela(LHCsqrts, mh, myVerbosity_);

    fL = f_L;
    if (f_L == 1.0) {
      a1 = 0.0;
    }
    
  }

  std::vector<RVecF> operator()(int _nLHEPart,
				RVecF _LHEPart_pt,
				RVecF _LHEPart_eta,
				RVecF _LHEPart_phi,
				RVecF _LHEPart_mass,
				RVecI _LHEPart_incomingpz,
				RVecI _LHEPart_pdgId,
				RVecI _LHEPart_status,
				RVecI _LHEPart_spin,
				RVecI _GenPart_genPartIdxMother,
				RVecI _GenPart_pdgId,
				RVecI _GenPart_status,
				RVecF _GenPart_pt,
				RVecF _GenPart_eta,
				RVecF _GenPart_phi,
				RVecF _GenPart_mass,
				float _Generator_x1,
				float _Generator_x2,
				int _Generator_id1,
				int _Generator_id2) {
    
    Generator_x1 = _Generator_x1;
    Generator_x2 = _Generator_x2;

    Generator_id1 = _Generator_id1;
    Generator_id2 = _Generator_id2;

    nLHEPart = _nLHEPart;
    LHEPart_pt = _LHEPart_pt;
    LHEPart_eta = _LHEPart_eta;
    LHEPart_phi = _LHEPart_phi;
    LHEPart_mass = _LHEPart_mass;
    LHEPart_incomingpz = _LHEPart_incomingpz;
    LHEPart_pdgId = _LHEPart_pdgId;
    LHEPart_status = _LHEPart_status;
    LHEPart_spin = _LHEPart_spin;

    GenPart_genPartIdxMother = _GenPart_genPartIdxMother;
    GenPart_pdgId = _GenPart_pdgId;
    GenPart_status = _GenPart_status;
    GenPart_pt = _GenPart_pt;
    GenPart_eta = _GenPart_eta;
    GenPart_phi = _GenPart_phi;
    GenPart_mass = _GenPart_mass;

    //std::cout << "" << std::endl;
    //std::cout << "New event ---------------------" << std::endl;
    //std::cout << "" << std::endl;
    
    std::vector<MELAParticle> melaParticlesList;

    std::vector<TLorentzVector> incoming;
    std::vector<int> incomingIDs;
    std::vector<TLorentzVector>	outgoing;
    std::vector<int> outgoingIDs;

    TLorentzVector Higgs;
    bool first = true;
    bool areLeptons = false;
    
    for (int p = 0; p < nLHEPart; p++){
      TLorentzVector tmp;
      if (LHEPart_status[p] == 1){ // outgoing

	//cout << "Is stable!" << endl;
	//cout << LHEPart_pdgId[p] << endl;
	//cout << LHEPart_mass[p] << endl;

	if (abs(LHEPart_pdgId[p])>=11 && abs(LHEPart_pdgId[p])<=16){
	  areLeptons = true;
	}
	
	tmp.SetPtEtaPhiM(LHEPart_pt[p], LHEPart_eta[p], LHEPart_phi[p], LHEPart_mass[p]);
	if (first && abs(LHEPart_pdgId[p])<17 && abs(LHEPart_pdgId[p])>10){
	  Higgs = tmp;
	  first = false;
	}else if (abs(LHEPart_pdgId[p])<17 && abs(LHEPart_pdgId[p])>10){
	  Higgs = Higgs + tmp;
	}
	outgoing.push_back(tmp);
	outgoingIDs.push_back(LHEPart_pdgId[p]);
      }else if(LHEPart_status[p] == -1){

	//cout <<	"Is unstable!" << endl;
	//cout <<	LHEPart_pdgId[p] << endl;
	//cout << LHEPart_incomingpz[p] << endl;

	tmp.SetPxPyPzE(0.0, 0.0, LHEPart_incomingpz[p], abs(LHEPart_incomingpz[p]));
        incoming.push_back(tmp);
        incomingIDs.push_back(LHEPart_pdgId[p]);
      }      
    }
    
    SimpleParticle_t tempPart;
    MELAParticle tempMela;

    for (int idx=0; idx<outgoing.size(); idx++){
      tempPart = SimpleParticle_t(outgoingIDs[idx], outgoing[idx]);
      tempMela = MELAParticle(tempPart.first, tempPart.second);
      tempMela.setGenStatus(1);
      melaParticlesList.push_back(tempMela);
    }

    for	(int idx=0; idx<incoming.size(); idx++){
      tempPart = SimpleParticle_t(incomingIDs[idx], incoming[idx]);
      tempMela = MELAParticle(tempPart.first, tempPart.second);
      tempMela.setGenStatus(-1);
      melaParticlesList.push_back(tempMela);
    }
    
    MELAEvent* LHEEvent = new MELAEvent();
    std::vector<MELAParticle> writtenGenCands;
    std::vector<MELAParticle> writtenGenTopCands;
    bool ThereIsHiggs = false;

    for (auto& part : melaParticlesList){
      if (part.genStatus == -1){
	LHEEvent->addMother(&part);
      }else{
	if (PDGHelpers::isALepton(part.id)){
	  LHEEvent->addLepton(&part);
	}else if(PDGHelpers::isANeutrino(part.id)){
	  LHEEvent->addNeutrino(&part);
	}else if(PDGHelpers::isAKnownJet(part.id) && PDGHelpers::isATopQuark(part.id) == false){
	  LHEEvent->addJet(&part);
	}else if(PDGHelpers::isAPhoton(part.id)){
	  LHEEvent->addPhoton(&part);
	}else if(PDGHelpers::isAHiggs(part.id)){
	  writtenGenCands.push_back(part);
	  ThereIsHiggs = true;
	  if (VVMode==MELAEvent::UndecayedMode && (part.genStatus==1 || part.genStatus==2)){
	    LHEEvent->addIntermediate(&part);
	  }
	}else if(PDGHelpers::isATopQuark(part.id)){
	  writtenGenTopCands.push_back(part);
	  if (part.genStatus==1){
	    LHEEvent->addIntermediate(&part);
	  }
	}else{
	  cout << "FATAL: UNIDENTIFIED PARTICLE IN THE FINAL STATE" << endl;
	  cout << "ID: " << part.id << endl;
	}
      }
    }

    LHEEvent->constructTopCandidates();

    std::vector<MELATopCandidate_t*> matchedTops;
    for (auto& writtenGenTopCand : writtenGenTopCands){
      MELATopCandidate_t* tmpCand = TopComparators::matchATopToParticle(*LHEEvent, &writtenGenTopCand);
      if (tmpCand){
	matchedTops.push_back(tmpCand);
      }
    }
    for (auto& tmpCand : LHEEvent->getTopCandidates()){
      if (std::find(matchedTops.begin(), matchedTops.end(), tmpCand) != matchedTops.end()){
	tmpCand->setSelected(false);
      }
    }
    
    LHEEvent->constructVVCandidates(VVMode, VVDecayMode);
    LHEEvent->addVVCandidateAppendages();
    
    bool ThereIsCand = false;
    MELACandidate* genCand;

    bool doVBF = false;
    bool gluonsIn = false;
    std::vector<MELAParticle*> LHEmothers = LHEEvent->getMothers();
    for (auto& part : LHEmothers){
      if (part->id!=21)
	doVBF = true;
      if (part->id==21)
	gluonsIn = true;
    }
    
    if (ThereIsHiggs){
      for (auto& writtenGenCand : writtenGenCands){
        MELACandidate* tmpCand = HiggsComparators::matchAHiggsToParticle(*LHEEvent, &writtenGenCand);
	if (tmpCand){
	  if (!ThereIsCand){
	    genCand = tmpCand;
	    ThereIsCand = true;
	  }else{
	    genCand = HiggsComparators::candComparator(genCand, tmpCand, HiggsComparators::BestZ1ThenZ2ScSumPt, VVMode);
	  }
	}
      }
    }else{
      genCand = HiggsComparators::candidateSelector(*LHEEvent, HiggsComparators::BestZ1ThenZ2ScSumPt, VVMode);
    }
    
    // MAKE THE IvyAutoMELA CALL 
    //cout << "====================================================" << endl;
    //TUtil::PrintCandidateSummary(genCand);
    //cout << "====================================================" << endl;

    float m1 = -1.0;
    float m2;

    std::vector<MELAParticle*> LHEbosons = LHEEvent->getIntermediates();
    for (auto& part : LHEbosons){
      
      //std::cout << part->id << std::endl; ///// Check --------
      
      if (abs(part->id) == 23){
	if (m1 == -1.0)
	  m1 = part->m();
	else{
	  m2 = part->m();
	  if (m1<m2){
	    m2 = m1;
	    m1 = part->m();
	  }
	}
      }
    }

    mela->selfDHzzcoupl[0/1][0][0]=1;
    mela->selfDHwwcoupl[0/1][0][0]=1;
    mela->selfDHggcoupl[0/1][0][0]=1;    
    mela->setCandidateDecayMode(TVar::CandidateDecay_WW);
    mela->setCurrentCandidate(genCand);

    float qH;
    float mV1;
    float mV2;
    float costheta1;
    float costheta2;
    float Phi;
    float costhetastar;
    float Phi1;

    mela->setProcess(TVar::HSMHiggs, TVar::JHUGen, TVar::ZZINDEPENDENT);
    mela->setMelaHiggsMassWidth(125.0, 0.00407, 0);
    mela->computeDecayAngles(qH, mV1, mV2, costheta1, costheta2, Phi, costhetastar, Phi1);

    float genMass;
    genMass = (float)genCand->m();

    /////////
    ///////// PREPARE OUTPUT FOR MELAcalc.py
    /////////
    
    RVecF LHEDaughterId;
    RVecF LHEDaughterPt;
    RVecF LHEDaughterEta;
    RVecF LHEDaughterPhi;
    RVecF LHEDaughterMass;

    RVecF LHEAssociatedParticlePt;
    RVecF LHEAssociatedParticleEta;
    RVecF LHEAssociatedParticlePhi;
    RVecF LHEAssociatedParticleMass;
    RVecF LHEAssociatedParticleId;

    RVecF LHEMotherId;
    RVecF LHEMotherPz;
    RVecF LHEMotherE;

    std::vector<MELAParticle*> LHEDaughters = genCand->getSortedDaughters();
    for (auto& part : LHEDaughters){
      LHEDaughterId.push_back((float)part->id);
      LHEDaughterPt.push_back(part->pt());
      LHEDaughterEta.push_back(part->eta());
      LHEDaughterPhi.push_back(part->phi());
      LHEDaughterMass.push_back(part->m());
    }

    std::vector<MELAParticle*> LHEAssociated = genCand->getAssociatedJets();
    for (auto& part : LHEAssociated){
      LHEAssociatedParticleId.push_back((float)part->id);
      LHEAssociatedParticlePt.push_back(part->pt());
      LHEAssociatedParticleEta.push_back(part->eta());
      LHEAssociatedParticlePhi.push_back(part->phi());
      LHEAssociatedParticleMass.push_back(part->m());
    }

    std::vector<MELAParticle*> LHEMothers = LHEEvent->getMothers();
    for (auto& part : LHEMothers){
      LHEMotherId.push_back((float)part->id);
      LHEMotherPz.push_back(part->z());
      LHEMotherE.push_back(part->t());
    }

    RVecF otherVars = {qH, mV1, mV2, costheta1, costheta2, Phi, costhetastar, Phi1, genMass};

    std::vector<RVecF> results_to_return = {
      LHEDaughterId,
      LHEDaughterPt,
      LHEDaughterEta,
      LHEDaughterPhi,
      LHEDaughterMass,
      LHEAssociatedParticlePt,
      LHEAssociatedParticleEta,
      LHEAssociatedParticlePhi,
      LHEAssociatedParticleMass,
      LHEAssociatedParticleId,
      LHEMotherId,
      LHEMotherPz,
      LHEMotherE,
      otherVars
    };

    return results_to_return;
  
    /**
    doVBF = false;
    if (!doVBF){ // skip VBF computation

      //cout << "Skip VBF Interference!" << endl;
      
      //cout << "----- RESULTS -----" << endl;
      //cout << "|ME| ggH Signal ->" << GG_SIG_kappaTopBot_1_ghz1_1_MCFM << endl;
      //cout << "|ME| ggH SBI    ->" << GG_BSI_kappaTopBot_1_ghz1_1_MCFM << endl;
      //cout << "|ME| ggH Bkg    ->" << GG_BKG_MCFM << endl;
      //cout << "|ME| VBF Signal ->" << JJEW_SIG_ghv1_1_MCFM << endl;
      //cout << "|ME| VBF SBI    ->" << JJEW_BSI_ghv1_1_MCFM << endl;
      //cout << "|ME| VBF Bkg    ->" << JJEW_BKG_MCFM << endl;

      genMass = (float)genCand->m();

      if (std::isnan(GG_SIG_azh1_azh2_JHUGen)) GG_SIG_azh1_azh2_JHUGen = 0.0;
      
      RVecF results = {GG_SIG_azh1_azh2_JHUGen, qH, m1, m2, costheta1, costheta2, Phi, costhetastar, Phi1, genMass};
      mela->resetInputEvent();
      return results;
    }
    
    //// Some tricks for VBF production
    TVar::Production candScheme=TVar::JJVBF;
    MELACandidateRecaster* recaster =  new MELACandidateRecaster(candScheme);
    MELACandidate* genCand_VBF;
      
    std::vector<MELAParticle*> associated_MELA = genCand->getAssociatedJets();
    bool gluonsOut = false;
    for (auto& part: associated_MELA){
      if (part->id == 21)
	gluonsOut = true;
    }    

    // Check gluons
    if (gluonsOut && !gluonsIn){
      // cout << "DO RECAST!" << endl;

      recaster->copyCandidate(genCand, genCand_VBF); 
      recaster->reduceJJtoQuarks(genCand_VBF);
      mela->setCurrentCandidate(genCand_VBF);
      
    }else{
      mela->setCurrentCandidate(genCand);
    }
    
    mela->setProcess(TVar::SelfDefine_spin0,TVar::JHUGen, TVar::JJEW);
    mela->setMelaHiggsMassWidth(125.0, 0.00407, 0);
    mela->computeP(GG_SIG_azh1_azh2_JHUGen, true);
    mela->computeDecayAngles(qH, m1, m2, costheta1, costheta2, Phi, costhetastar, Phi1);

    genMass = (float)genCand->m();

    if (std::isnan(GG_SIG_azh1_azh2_JHUGen)) GG_SIG_azh1_azh2_JHUGen = 0.0;

    RVecF results = {GG_SIG_azh1_azh2_JHUGen, qH, m1, m2, costheta1, costheta2, Phi, costhetastar, Phi1, genMass};
    mela->resetInputEvent();
    return results;
   
    **/
  }
};
